cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,243 @@
+ </struct>
+```
+
+Since the DDS Event Format currently supports only three types of data payloads, these are defined within the io::cloudevents DDS module by the folowing enumeration and union:
+
+```xml
+ <enum name="DataKind">
+   <enumerator name="BINARY"/>
+   <enumerator name="TEXT"/>
+   <enumerator name="JSON"/>
+ </enum>
+	  
+ <union name="Data" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::DataKind"/>
+   <case>
+      <caseDiscriminator value="(io::cloudevents::BINARY)"/>
+       <member name="binary_data" type="byte" sequenceMaxLength="100"/>
+    </case>
+    <case>
+      <caseDiscriminator value="(io::cloudevents::JSON)"/>
+       <member name="json_dds_data" type="string" stringMaxLength="255"/>
+    </case>
+    <case>
+      <caseDiscriminator value="(io::cloudevents::TEXT)"/>
+      <member name="text_data" type="string" stringMaxLength="255"/>
+    </case>
+ </union>
+```
+
+Finally, the complete [DDS Event Format][dds-event-format] structure is:
+
+```xml
+
+<struct name="Event" extensibility="mutable">
+   <member name="headers" type="nonBasic" nonBasicTypeName="io::cloudevents::Headers"/>
+   <member name="id" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string"/>
+   <member name="source" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_uri_reference"/>
+   <member name="specversion" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string"/>
+   <member name="type" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string"/>
+   <member name="datacontenttype" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string" optional="true"/>
+   <member name="datacontentencoding" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string" optional="true"/>
+   <member name="dataschema" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_uri" optional="true"/>
+   <member name="subject" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string" optional="true"/>
+   <member name="time" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_timestamp" optional="true"/>	    
+   <member name="extension" type="nonBasic" nonBasicTypeName="io::cloudevents::Attributes" optional="true"/>
+   <member name="datakey" type="nonBasic" nonBasicTypeName="io::cloudevents::ce_string" key="true"/>
+   <member name="body" type="nonBasic" nonBasicTypeName="io::cloudevents::Data" optional="true"/>
+</struct>
+```
+
+## 3 Data
+
+Before encoding, the DDS serializer MUST first determine the runtime data type
+of the content. This can be determined by examining by consulting the `datacontenttype`
+and `datacontentencoding` attributes.
+
+If the implementation determines that the type of the data is binary, the value
+MUST be stored in the `body` field using the `bytes` type.
+
+For other types, the implementation MUST translate the data value into a text or JSON
+representation using the union type described for the message body.
+
+## 4 Examples
```

Comment:
I wonder if it would be useful to DDS folks to see a full XML CE serialization, in addition to the table below?

Reply:
The long story: The XML format here is what is used to define what is required/optional in the DDS CE message. This format information is then provided to a DDS code generator that defines the message classes (code in C, C++, Java, Javascript, other supported languages) with the required attributes. The data values are populated in code into the message classes and the data is actually sent over the wire in a CDR (Common Data Representation) format which is then decoded by the DDS message receiver on the other side. So the XML is not the serialization format itself, it is simply the information used to configure the CDR serialization/deserialization process.

This was obviously not clearly written down by me. I could use some advice though on how to best put state this information in the document. Could we have a discussion on this topic? I will see if I can get on the agenda at tomorrow's meeting.